File: code/python/剑指offer37.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return []
        res = []
        queue = [root]
        while queue:
            cur = queue[0]
            queue = queue[1:]
            if cur is not None:
                res.append(cur.val)
                queue.append(cur.left)
                queue.append(cur.right)
            else:
                # 不添加空节点的左右节点
                res.append(None)
        for i in range(len(res) - 1, -1, -1):
            if res[i] is not None:
                break
        res = res[:i + 1]
        return res

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        # print(data)
        root = TreeNode(data[0])
        queue = [root]
        index = 0
        # 按照层序遍历构建二叉树
        while index < len(data):
            cur = queue[0]
            queue = queue[1:]
            index += 1
            if index < len(data):
                if data[index] is not None:
                    cur.left = TreeNode(data[index])
                    queue.append(cur.left)
                else:
                    cur.left = None
            index += 1
            if index < len(data):
                if data[index] is not None:
                    cur.right = TreeNode(data[index])
                    queue.append(cur.right)
                else:
                    cur.right = None
        return root

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: code/java/剑指offer37.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root==null) return "[]";
        Queue<TreeNode> queue = new LinkedList<>();
        StringBuilder res = new StringBuilder("[");
        queue.offer(root);
        while(queue.size()>0){
            TreeNode cur = queue.remove();
            if(cur!=null){
                res.append(cur.val + ",");
                queue.offer(cur.left);
                queue.offer(cur.right);
            }else{
                res.append("null,");
            }
        }
        res.deleteCharAt(res.length() - 1);
        res.append("]");
        return res.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.equals("[]")) return null;
        String[] values = data.substring(1,data.length()-1).split(",");
        TreeNode root = new TreeNode(Integer.parseInt(values[0]));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;
        while(queue.size()>0){
            TreeNode node = queue.remove();
            if(!values[i].equals("null")){
                node.left = new TreeNode(Integer.parseInt(values[i]));
                queue.add(node.left);
            }
            i+=1;
            if(!values[i].equals("null")){
                node.right = new TreeNode(Integer.parseInt(values[i]));
                queue.add(node.right);
            }
            i+=1;
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));